pkg/reconciler/managed/reconciler.go

```diff
@@ -721,30 +804,49 @@ func (r *Reconciler) Reconcile(_ context.Context, req reconcile.Request) (reconc
+
+			// We handle annotations specially here because it's
+			// critical that they are persisted to the API server.
+			// If we don't remove the external-create-pending
+			// annotation the reconciler will refuse to proceed,
+			// because it won't know whether or not it created an
+			// external resource.
+			meta.SetExternalCreateFailed(managed, time.Now())
+			if err := r.managed.UpdateCriticalAnnotations(ctx, managed); err != nil {
+				log.Debug(errUpdateManagedAnnotations, "error", err)
+				record.Event(managed, event.Warning(reasonCannotUpdateManaged, errors.Wrap(err, errUpdateManagedAnnotations)))
+
+				// We only log and emit an event here rather
+				// than setting a status condition and returning
+				// early because presumably it's more useful to
+				// set our status condition to the reason the
+				// create failed.
+			}
+
 			managed.SetConditions(xpv1.ReconcileError(errors.Wrap(err, errReconcileCreate)))
 			return reconcile.Result{Requeue: true}, errors.Wrap(r.client.Status().Update(ctx, managed), errUpdateManagedStatus)
 		}
 
-		if creation.ExternalNameAssigned {
-			en := meta.GetExternalName(managed)
-			// We will retry in all cases where the error comes from the api-server.
-			// At one point, context deadline will be exceeded and we'll get out
-			// of the loop. In that case, we warn the user that the external resource
-			// might be leaked.
-			err := retry.OnError(retry.DefaultRetry, resource.IsAPIError, func() error {
-				nn := types.NamespacedName{Name: managed.GetName()}
-				if err := r.client.Get(ctx, nn, managed); err != nil {
-					return err
-				}
-				meta.SetExternalName(managed, en)
-				return r.client.Update(ctx, managed)
-			})
-			if err != nil {
-				log.Debug("Cannot update managed resource", "error", err)
-				record.Event(managed, event.Warning(reasonCannotUpdateManaged, errors.Wrap(err, errUpdateManagedAfterCreate)))
-				managed.SetConditions(xpv1.ReconcileError(errors.Wrap(err, errUpdateManagedAfterCreate)))
-				return reconcile.Result{Requeue: true}, errors.Wrap(r.client.Status().Update(ctx, managed), errUpdateManagedStatus)
-			}
+		// In some cases our external-name may be set by Create above.
+		log = log.WithValues("external-name", meta.GetExternalName(managed))
+		record = r.record.WithAnnotations("external-name", meta.GetExternalName(managed))
+
+		// We handle annotations specially here because it's critical
+		// that they are persisted to the API server. If we don't remove
+		// the external-create-pending annotation the reconciler will
+		// refuse to proceed, because it won't know whether or not it
+		// created an external resource. This is important in cases
+		// where we must record an external-name annotation set by the
+		// Create call. Any other changes made during Create will be
+		// reverted when annotations are updated; at the time of writing
+		// Create implementations are advised not to alter status, but
```

Comment:
We could possibly do something like `managed.GetStatus().DeepCopy` and after the critical update, `managed.SetStatus(copied)` but that'd require change in `Managed` interface and the code generator, so maybe in the future. But I'll open an issue just to have a place to see how big of a problem this will be.

Reply:
Yeah @turkenh suggested the same thing on a previous iteration of this PR. I'd prefer to avoid it unless there's a use case for modifying spec and/or status during `Create`.